### PR TITLE
`Exam mode`: Send problem statement notifications only after the start of an exam

### PR DESCRIPTION
**Pull request by edkaya**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

This feature is a part of the new approach for exam notifications discussed in #9121 and also fixes the issue. Only after the exam started, the server should produce the actual exam live event for problem statement changes and display it all students who are affected.

### Description

Added condition for creating exam live events for problem statement changes
Changed from exercise title to exercise group title in problem statement update notification
Adapted java tests accordingly by overloading `addExerciseGroupWithExamAndCourse` and adding a boolean parameter to `addCourseExamExerciseGroupWithOneFileUploadExercise` methods to ensure that exam live events are tested with an already created exam.
 
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam

1. Log in to Artemis as an instructor
2. Create an Exam 
3. Make some changes in problem statement
4. Log in as student
5. Go to the exam 
6. Verify that there are no notifications displayed before the start date of the exam
7. Wait for the exam start date and start the exam
8. Verify that problem statement has changed after the start of the exam
9. Log in to Artemis as an instructor
10. Make some changes in problem statement again
11. Log in as student
12. Verify that you see the changes and notifications too

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam 

1. Log in to Artemis
2. Participate in the exam as a student
3. Make sure that you can see all notifications made during the exam.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2
#### Exam Mode Test
- [x] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
  - Enhanced notification system for exercise changes to ensure students are only informed if the exam is imminent.
  - Improved clarity in event notifications by deriving exercise names from their respective groups.
  - Added support for filtering event notifications based on the exam start date.

- **Bug Fixes**
  - Streamlined logic for notifying students about working time updates during exams, ensuring timely communication.

- **Tests**
  - Updated test cases to more accurately simulate real-world scenarios regarding exam visibility and timing.
  - Enhanced flexibility in test setups with new parameters for exercise group creation, allowing for better configuration control.

### CODE REVIEW

**Review by EneaGore — COMMENTED** [review]
So it works exactly how the steps suggest. But there may be an unintented side-effect. I had one user just wait, and the other one start early and wait for the start. The one who started early was shown the not updated time, and it updated after leaving and entering the exam again.
this cannot be good

**Comment by edkaya** [comment]
> So it works exactly how the steps suggest. But there may be an unintented side-effect. I had one user just wait, and the other one start early and wait for the start. The one who started early was shown the not updated time, and it updated after leaving and entering the exam again. this cannot be good

You are right, time needs to be changed in live and not with a refresh. I looked into that and I think it will require additional changes on client side too. Since displaying time updates is not the main issue discussed in #9121 and therefore not urgent, I would suggest doing this in a follow-up PR. Thanks for noticing 👍

**Review by N0W0RK — DISMISSED** [review]
Code 👍

**Review by EneaGore — COMMENTED** [review]
Tested on TS6. The user does not see the exercise change notification anymore so nothing can be leaked. However, the same issue persist even for exercise changes. User on the left was on start exam countdown while the changes where made and he has no way to know that something has changed. I am not 100% sure if this is linked specifically to this PR but now with the Working Time notification the time is shown correctly for both. The `createAndSendProblemStatementUpdateEvent` seems to do more than just send a notification. Its somewhat of a niche and unlikely scenario but idk.
rabbitNotUpdate

**Review by JohannesStoehr — DISMISSED** [review]
Code

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java [hunk 1]
```diff
@@ -777,7 +778,11 @@ public void notifyAboutExerciseChanges(Exercise originalExercise, Exercise updat
         }
         else if (originalExercise.isExamExercise() && !StringUtils.equals(originalExercise.getProblemStatement(), updatedExercise.getProblemStatement())) {
             User instructor = userRepository.getUser();
-            this.examLiveEventsService.createAndSendProblemStatementUpdateEvent(updatedExercise, notificationText, instructor);
+            Exam exam = originalExercise.getExam();
+            // notify the student only if the exam is started
+            if (exam.isStarted()) {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:783]: This check should be before comparing the problem statements as this comparison is more complex.

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by N0W0RK — APPROVED** [review]
Code

**Review by EneaGore — APPROVED** [review]
Tested on TS6. Works as expected. Thank you for fixing the refresh thing.

**Review by anian03 — APPROVED** [review]
Tested on TS6, everything works as described. No notifications about changes to the problem statement were sent before the start of the exam, and during the exam notifications were received. Problem statements were updated correctly

**Review by sarpsahinalp — APPROVED** [review]
Tested on ts6. Works as expected

**Review by pzdr7 — APPROVED** [review]
The code looks good 👍

**Review by krusche — APPROVED** [review]

**Pull request merged**